Inclusão de uma entidade auditável

damews opened the pull request:
O que foi feito:
Inclusão de uma entidade auditável.
Utilização da auditoria na _dataRegistro_ do Pet (a classe extende a auditável)

**Classe auditável adicionada.**
Agora é possível saber quando a entidade foi criada.

**Configurações para implementação adicionadas**
Auditoria começa quando a aplicação inicia.
Links de auditoria para as classes que a utilizam

**Mudanças nos DTOs**
O PetDTO não recebe mais uma data de registro, pois a data é gerada automaticamente quando salva no banco. [Removido a propriedade]

_Adicionado_: +PetRespostaDTO
É o PetDTO de antigo (contém a data de registro como propriedade). Agora todas as respostas para o front-end serão com este modelo.

**Mudanças no controller/service***
Agora os retornos são com o PetRespostaDTO

damews committed:
Inclusão de uma entidade auditável

Classe auditável adicionada
Configurações para implementação adicionadas

damews committed:
Arrumando Entidade Auditável

Agora é somente para o campo dataRegistro da classe Pet
-Arrumando tests, builders

damews committed:
Circle Ci

Testes de localização por raio dando erro. (Comentado)
Teste mapper: removido getDataRegistro (o campo é gerado auto)

nefasta commented on src/test/java/br/com/academiadev/thunderpets/controller/PetControllerTest.java:
```diff
@@ -302,8 +303,8 @@ public void dadoPets_quandoBuscoPorRaioDeDistancia3OrdenacaoNomeASC_entaoBuscarP
-                .andExpect(jsonPath("$.content[0].nome", is("Carijó")))
-                .andExpect(jsonPath("$.content[0].localizacao.descricao", is("R. Dona Francisca - Centro, Joinville - SC, 89201-070")))
+//                .andExpect(jsonPath("$.content[0].nome", is("Carijó")))
+//                .andExpect(jsonPath("$.content[0].localizacao.descricao", is("R. Dona Francisca - Centro, Joinville - SC, 89201-070")))
```
Esse teste aqui a princípio passou no último [build](https://circleci.com/gh/academiadev-jlle/backend-thunderpets/184?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) do CircleCI, acho que dá para descomentar essas linhas que não haverá problemas. 😁

nefasta commented on src/test/java/br/com/academiadev/thunderpets/controller/PetControllerTest.java:
```diff
@@ -257,13 +258,13 @@ public void dadoPets_quandoBuscoPorCidadeJoinvilleOrdenacaoPadrao_entaoBuscarPet
-                .andExpect(jsonPath("$.content[1].localizacao.cidade", is("Joinville")))
-                .andExpect(jsonPath("$.content[1].localizacao.estado", is("Santa Catarina")));
+//        listaDePets.andExpect(status().isOk())
+//                .andExpect(jsonPath("$.content[0].nome", is("Pocotó")))
```
O teste falhou porque a busca por Pets por padrão é feita por dataRegistro DESC, conforme `PetController`:

> @ApiParam(value = "Direção da ordenação: ascendente ou descendente")
  @RequestParam(defaultValue = "DESC") Sort.Direction direcao,
@ApiParam(value = "Nome da coluna que será usada para a ordenação")
  @RequestParam(defaultValue = "dataRegistro") String campoOrdenacao,

E, quando eu criei os testes eu adicionei valores fixos ao atributo dataRegistro, conforme `PetDTOUtil`:

> .nome("Carijó")
.dataRegistro(LocalDate.of(2018, 8,4))

> .nome("Pocotó")
.dataRegistro(LocalDate.of(2018, 12,1))

Desculpeee, não havia lembrado da auditoria dos pets quando fiz os testes. My bad 😐

Mas entãooo, acho que dá pra remover esse atributo lá do builder, igual você fez nas classes `PetMapperTest` e `PetUtil`. Logo, como no teste em questão o pet Carijó é criado antes, é só inverter. Dá pra criar o pet Pocotó antes ou inverter os índices dos `jsonPath` esperados.

damews committed:
Arrumando requisições pull request

damews committed:
Alterações em DTOs

Agora possui um PetRespostaDTO, contendo a prop. dataRegistro.
O PetDTO (usado para salvar), não possui a dataRegistro que é gerada auto.

damews committed:
Circle Ci

rartner commented on src/main/java/br/com/academiadev/thunderpets/config/audit/SpringSecurityAuditorAware.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Optional;
+
+@Component
+public class SpringSecurityAuditorAware implements AuditorAware<String> {
```
Por que precisa dessa classe? Não lembro da gente utilizando isso nos exemplos

LiserLine replied:
Eu coloquei nos primeiros testes... Isso servia pra pegar o cara que estava atualizando aquela entidade. Servia basicamente pra atribuir a prop. createdBy/updatedBy.
**Removido**

rartner commented on src/main/java/br/com/academiadev/thunderpets/model/EntidadeAuditavel.java:
```diff
@@ -0,0 +1,27 @@
+
+    @CreatedDate
+    @Column(updatable = false)
+    private LocalDate dataRegistro;
```
Tava pensando aqui, talvez seja melhor colocar um LocalDateTime. Vai ajudar na hora da ordenação e tal

LiserLine replied:
Colocado!

rartner commented on src/main/java/br/com/academiadev/thunderpets/model/Pet.java:
```diff
@@ -60,4 +54,20 @@
     private boolean ativo = true;
+
+    @Builder
+    public Pet(LocalDate dataRegistro, String nome, @NotNull String descricao, LocalDate dataAchado, @NotNull Especie especie, @NotNull Porte porte, @NotNull Sexo sexo, @NotNull Status status, @NotNull Idade idade, Usuario usuario, Localizacao localizacao, boolean ativo) {
```
Nunca vi um construtor com o `@Builder`. Pra quê serve?

LiserLine replied:
Se colocado o @Builder na classe filho e na classe pai o método builder() não funciona, pois ele não consegue encontrar a variável do pai. Uma das formas é colocar o construtor na classe filho colocando o @Builder na mesma e chamando a variável do pai com o super(). A outra forma é utilizar o @SuperBuilder, porém eu não consegui com que funcionasse.

rartner commented on src/main/java/br/com/academiadev/thunderpets/model/Usuario.java:
```diff
@@ -43,6 +43,7 @@
     @Lob
     private byte[] foto;
 
+    @Builder.Default
```
Pra quê serve isso, migo?

LiserLine replied:
O log dava um _warning_ dizendo que se não colocar a anotação o valor padrão que foi atribuido não iria ser utilizado, deixando a prop. sem valor default.

rartner commented on src/main/java/br/com/academiadev/thunderpets/service/impl/PetServiceImpl.java:
```diff
@@ -100,23 +100,21 @@ public PetServiceImpl(PetRepository petRepository,
                 throw new ErroAoProcessarException("Para buscas por raio de distância é necessário informar a lagitude e longitude do usuário atual.");
```
Não foi você que alterou, mas tá escrito `lagitude`. Consegue arrumar pra gente? :heart:

LiserLine replied:
👍

nefasta replied:
Ops, hehehehe. Lagitude e longitude!

damews committed:
Pedidos do PR

Removido classes inutilizáveis
Trocando LocalDate para LocalDateTime na EntidadeAuditável
Arrumando as funções para que utilizem o novo tipo.

rartner committed:
Merge branch 'release/4' into feature/36
